File: day5/part1/d5p1.js
const fs = require('fs');

fs.readFile("realinput.txt", 'utf8', (err, data) => {
    if (err) {
        console.error('Error reading the file:', err);
        return;
    }

    const inputText = data
    const inputLines = inputText.split("\r\n");

    function mapNavigator(startValue, destinationRangeStart, sourceRangeStart, rangeLength) {
        let outputArray = [];
        let mappedValue = startValue;
        outputArray[0] = mappedValue;
        outputArray[1] = false;
        if ((startValue >= sourceRangeStart) && (startValue <= (sourceRangeStart + rangeLength - 1))) {
            rangeStartDifference = Math.abs(sourceRangeStart - destinationRangeStart)
            mappedValue = (sourceRangeStart > destinationRangeStart) ? startValue - rangeStartDifference : startValue + rangeStartDifference
            outputArray[0] = mappedValue
            outputArray[1] = true;
        }
        return outputArray;
    }

    const seeds = inputLines[0].split(" ").slice(1).map(element => parseInt(element))
    let maps = inputText.split(":").slice(2).map(element => element.split("\r\n"))
    maps = maps.map(element => element.filter(subelement => Number.isInteger(parseInt(subelement[0]))))
    maps = maps.map(element => element.map(subelement => subelement.split(" ")))
    maps = maps.map(element => element.map(subelement => subelement.map(subsubelement => parseInt(subsubelement))))

    let lowestLocation;
    for (let k = 0; k < seeds.length; k++) {
        let XToY = seeds[k]
        let mapNavigatorOutput = [];
        maps.forEach(map => {
            for (let i = 0; i < map.length; i++) {
                mapNavigatorOutput = mapNavigator(XToY, map[i][0], map[i][1], map[i][2])
                XToY = mapNavigatorOutput[0]
                if (mapNavigatorOutput[1]) break
            }
        })
        if (lowestLocation == undefined) lowestLocation = XToY
        if (XToY < lowestLocation) lowestLocation = XToY
    }

    console.log(lowestLocation)

})

File: day5/part1/d5p1.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>

std::vector<long long> mapNavigator(long long startValue, long long destinationRangeStart, long long sourceRangeStart, long long rangeLength)
{
    std::vector<long long> outputVector(2);
    long long mappedValue = startValue;
    outputVector[0] = mappedValue;
    outputVector[1] = false;

    if ((startValue >= sourceRangeStart) && (startValue <= (sourceRangeStart + rangeLength - 1)))
    {
        long long rangeStartDifference = std::abs(sourceRangeStart - destinationRangeStart);
        mappedValue = (sourceRangeStart > destinationRangeStart) ? (startValue - rangeStartDifference) : (startValue + rangeStartDifference);
        outputVector[0] = mappedValue;
        outputVector[1] = true;
    }

    return outputVector;
}

int main()
{
    std::ifstream inputFile("realinput.txt");
    if (!inputFile.is_open())
    {
        std::cerr << "Error opening the file" << std::endl;
        return 1;
    }

    std::vector<std::string> inputLines;
    std::string line;
    while (std::getline(inputFile, line))
    {
        inputLines.push_back(line);
    }
    inputFile.close();

    std::istringstream iss(inputLines[0]);

    std::string ignoreString;
    iss >> ignoreString;

    long long num;
    std::vector<long long> seeds;
    while (iss >> num)
    {
        seeds.push_back(num);
    }

    std::vector<std::vector<long long>> maps;
    long counter = -1;
    bool colonEncountered = false;
    bool emptyLineEncountered = false;
    for (long i = 2; i < inputLines.size(); ++i)
    {
        if (inputLines[i].find(':') != std::string::npos)
        {
            colonEncountered = true;
            counter++;
            i++;
        }

        if (inputLines[i].empty())
        {
            emptyLineEncountered = true;
        }

        if (colonEncountered && emptyLineEncountered)
        {
            colonEncountered = false;
            emptyLineEncountered = false;
        }
        else
        {
            if (colonEncountered && !emptyLineEncountered)
            {
                std::istringstream iss(inputLines[i]);
                long long num;
                if (counter >= maps.size())
                {
                    maps.push_back(std::vector<long long>());
                }

                while (iss >> num)
                {
                    maps[counter].push_back(num);
                }
            }
        }
    }

    long long lowestLocation = -1;
    for (int i = 0; i < seeds.size(); i++)
    {
        long long xToY = seeds[i];
        std::vector<long long> mapNavigatorOutput(2);
        for (int k = 0; k < maps.size(); k++)
        {
            for (int l = 0; l < maps[k].size() / 3; l++)
            {
                mapNavigatorOutput = mapNavigator(xToY, maps[k][0 + l * 3], maps[k][1 + l * 3], maps[k][2 + l * 3]);
                xToY = mapNavigatorOutput[0];
                if (mapNavigatorOutput[1])
                    break;
            }
        }
        if (lowestLocation == -1)
            lowestLocation = xToY;
        else if (xToY < lowestLocation)
            lowestLocation = xToY;
    }

    std::cout << lowestLocation;

    return 0;
}
